media-proxy/include/mesh/st2110tx.h

```diff
@@ -0,0 +1,153 @@
+  public:
+    ST2110Tx()
+    {
+        _kind = Kind::transmitter;
+
+        _handle = nullptr;
+        _ops = {0};
+        _transfer_size = 0;
+    };
+    ~ST2110Tx() { shutdown(_ctx); };
+
+  protected:
+    HANDLE _handle;
+    OPS _ops;
+    uint32_t _transfer_size;
+    context::Context _ctx;
+
+    virtual FRAME *get_frame(HANDLE) = 0;
+    virtual int put_frame(HANDLE, FRAME *) = 0;
+    virtual HANDLE create_session(mtl_handle, OPS *) = 0;
+    virtual int close_session(HANDLE) = 0;
+
+    Result on_establish(context::Context &ctx) override
+    {
+        _ctx = context::WithCancel(ctx);
+        _stop = false;
+
+        _handle = create_session(_st, &_ops);
+        if (!_handle) {
+            log::error("Failed to create session");
+            set_state(ctx, State::closed);
+            return set_result(Result::error_general_failure);
+        }
+
+        set_state(ctx, State::active);
+        return set_result(Result::success);
+    };
+
+    Result on_shutdown(context::Context &ctx) override
+    {
+        _ctx.cancel();
+
+        if (_handle) {
+            close_session(_handle);
+            _handle = nullptr;
+        }
+        set_state(ctx, State::closed);
+        return set_result(Result::success);
+    };
+
+    Result on_receive(context::Context &ctx, void *ptr, uint32_t sz, uint32_t &sent) override
+    {
+        sent = std::min(_transfer_size, sz);
+        // TODO: add error/warning if sent is different than _transfer_size
+
+        FRAME *frame = NULL;
+        do {
+            // Get empty buffer from MTL
+            frame = get_frame(_handle);
+            if (!frame) {
+                std::unique_lock lk(_mx);
+                _cv.wait(lk, _ctx.stop_token(), [this] { return _stop.load(); });
+                _stop = false;
+                if (_ctx.cancelled()) {
+                    break;
+                }
+            }
+        } while (!frame);
+
+        if (frame) {
+            // Copy data from emulated transmitter to MTL empty buffer
+            mtl_memcpy(get_frame_data_ptr(frame), ptr, sent);
+            // Return full buffer to MTL
+            put_frame(_handle, frame);
+        } else {
+            sent = 0;
+            return set_result(Result::error_shutdown);
+        }
+        return set_result(Result::success);
+    };
+
+  private:
+};
+
+class ST2110_20Tx : public ST2110Tx<st_frame, st20p_tx_handle, st20p_tx_ops> {
+  public:
+    ST2110_20Tx();
+    ~ST2110_20Tx();
+
+    Result configure(context::Context &ctx, const std::string &dev_port,
+                     const MeshConfig_ST2110 &cfg_st2110, const MeshConfig_Video &cfg_video);
+
+  protected:
+    st_frame *get_frame(st20p_tx_handle h) override;
+    int put_frame(st20p_tx_handle h, st_frame *f) override;
+    st20p_tx_handle create_session(mtl_handle h, st20p_tx_ops *o) override;
+    int close_session(st20p_tx_handle h) override;
+
+  private:
```

Comment:
```suggestion
```

Reply:
Fixed in e69185dab06ea1983303046f28572d6baf467706